server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java

```diff
@@ -131,6 +133,41 @@ public void writeBlob(String blobName, InputStream inputStream, long blobSize) t
         IOUtils.fsync(path, true);
     }
 
+    @Override
+    public void writeBlobAtomic(final String blobName, final InputStream inputStream, final long blobSize) throws IOException {
+        final Path tempBlobPath = path.resolve(tempBlobName(blobName));
+        try {
+            try (OutputStream outputStream = Files.newOutputStream(tempBlobPath, StandardOpenOption.CREATE_NEW)) {
+                Streams.copy(inputStream, outputStream);
+            }
+            IOUtils.fsync(tempBlobPath, false);
+
+            final Path blobPath = path.resolve(blobName);
+            // If the target file exists then Files.move() behaviour is implementation specific
+            // the existing file might be replaced or this method fails by throwing an IOException.
+            if (Files.exists(blobPath)) {
+                throw new FileAlreadyExistsException("blob [" + blobPath + "] already exists, cannot overwrite");
+            }
+            Files.move(tempBlobPath, blobPath, StandardCopyOption.ATOMIC_MOVE);
+        } finally {
+            IOUtils.deleteFilesIgnoringExceptions(tempBlobPath);
```

Comment:
Should we really ignore all exceptions here? This means we can up with lingering files without any warnings.
I preferred the previous approach of:

```
        } catch (IOException ex) {		
            // temporary blob creation or move failed - try cleaning up		
            try {		
                snapshotsBlobContainer.deleteBlobIgnoringIfNotExists(tempBlobName);		
            } catch (IOException e) {		
                ex.addSuppressed(e);		
            }		
            throw ex;		
          }
```

Reply:
Right, good point